Look at integrating "pylint" into the pipeline

In #137 I noticed a problem that Flake8, et al., had not detected. [Pylint](https://pylint.readthedocs.io/en/stable/) is able to detect that particular problem:

```
src/guiguts/maintext.py:422:24: W0622: Redefining built-in 'range' (redefined-builtin)
```

However, it reports much more, and at least some of that is noise. I'll look at integrating pylint and also suppressing some of its checks that we don't want. Sample output below.

```
❯ poetry run pylint ./src
************* Module guiguts.maintext
src/guiguts/maintext.py:717:0: C0301: Line too long (105/100) (line-too-long)
src/guiguts/maintext.py:27:0: R0903: Too few public methods (0/2) (too-few-public-methods)
src/guiguts/maintext.py:40:0: R0901: Too many ancestors (9/7) (too-many-ancestors)
src/guiguts/maintext.py:65:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/maintext.py:135:8: C0103: Attribute name "modifiedCallbacks" doesn't conform to snake_case naming style (invalid-name)
src/guiguts/maintext.py:81:0: R0901: Too many ancestors (9/7) (too-many-ancestors)
src/guiguts/maintext.py:81:0: R0902: Too many instance attributes (8/7) (too-many-instance-attributes)
src/guiguts/maintext.py:159:35: C0201: Consider iterating the dictionary directly instead of calling .keys() (consider-iterating-dictionary)
src/guiguts/maintext.py:215:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/maintext.py:268:28: W0108: Lambda may not be necessary (unnecessary-lambda)
src/guiguts/maintext.py:269:28: W0108: Lambda may not be necessary (unnecessary-lambda)
src/guiguts/maintext.py:283:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/maintext.py:375:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/maintext.py:380:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/maintext.py:385:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/maintext.py:422:24: W0622: Redefining built-in 'range' (redefined-builtin)
src/guiguts/maintext.py:466:12: W0622: Redefining built-in 'range' (redefined-builtin)
src/guiguts/maintext.py:479:4: R0914: Too many local variables (20/15) (too-many-locals)
src/guiguts/maintext.py:538:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/maintext.py:545:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/maintext.py:552:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/maintext.py:653:4: R0913: Too many arguments (7/5) (too-many-arguments)
src/guiguts/maintext.py:699:4: R0913: Too many arguments (6/5) (too-many-arguments)
src/guiguts/maintext.py:702:8: W0622: Redefining built-in 'range' (redefined-builtin)
src/guiguts/maintext.py:81:0: R0904: Too many public methods (48/20) (too-many-public-methods)
src/guiguts/maintext.py:745:0: C0103: Constant name "_single_widget" doesn't conform to UPPER_CASE naming style (invalid-name)
src/guiguts/maintext.py:750:4: W0603: Using the global statement (global-statement)
src/guiguts/maintext.py:6:0: C0411: standard import "import tkinter as tk" should be placed before "import regex as re" (wrong-import-order)
src/guiguts/maintext.py:7:0: C0411: standard import "from tkinter import ttk" should be placed before "import regex as re" (wrong-import-order)
src/guiguts/maintext.py:8:0: C0411: standard import "from tkinter import font as tk_font" should be placed before "import regex as re" (wrong-import-order)
src/guiguts/maintext.py:10:0: C0411: standard import "from typing import Any, Callable, Optional" should be placed before "import regex as re" (wrong-import-order)
************* Module guiguts.checkers
src/guiguts/checkers.py:23:34: W0622: Redefining built-in 'range' (redefined-builtin)
src/guiguts/checkers.py:15:0: R0903: Too few public methods (0/2) (too-few-public-methods)
src/guiguts/checkers.py:64:8: W0622: Redefining built-in 'range' (redefined-builtin)
************* Module guiguts.preferences
src/guiguts/preferences.py:133:13: W1514: Using open without explicitly specifying an encoding (unspecified-encoding)
src/guiguts/preferences.py:146:8: C0206: Consider iterating with .items() (consider-using-dict-items)
src/guiguts/preferences.py:146:19: C0201: Consider iterating the dictionary directly instead of calling .keys() (consider-iterating-dictionary)
************* Module guiguts.mainwindow
src/guiguts/mainwindow.py:684:0: C0301: Line too long (107/100) (line-too-long)
src/guiguts/mainwindow.py:77:29: W0613: Unused argument 'event' (unused-argument)
src/guiguts/mainwindow.py:141:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/mainwindow.py:163:4: R0913: Too many arguments (6/5) (too-many-arguments)
src/guiguts/mainwindow.py:193:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/mainwindow.py:233:0: R0902: Too many instance attributes (11/7) (too-many-instance-attributes)
src/guiguts/mainwindow.py:339:25: W0613: Unused argument 'event' (unused-argument)
src/guiguts/mainwindow.py:430:12: W0201: Attribute 'width' defined outside __init__ (attribute-defined-outside-init)
src/guiguts/mainwindow.py:430:24: W0201: Attribute 'height' defined outside __init__ (attribute-defined-outside-init)
src/guiguts/mainwindow.py:456:0: R0901: Too many ancestors (8/7) (too-many-ancestors)
src/guiguts/mainwindow.py:520:0: R0901: Too many ancestors (9/7) (too-many-ancestors)
src/guiguts/mainwindow.py:678:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/mainwindow.py:689:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/mainwindow.py:9:0: C0411: standard import "import time" should be placed before "from PIL import Image, ImageTk" (wrong-import-order)
src/guiguts/mainwindow.py:10:0: C0411: standard import "import traceback" should be placed before "from PIL import Image, ImageTk" (wrong-import-order)
src/guiguts/mainwindow.py:11:0: C0411: standard import "import tkinter as tk" should be placed before "from PIL import Image, ImageTk" (wrong-import-order)
src/guiguts/mainwindow.py:12:0: C0411: standard import "from tkinter import ttk, messagebox" should be placed before "from PIL import Image, ImageTk" (wrong-import-order)
src/guiguts/mainwindow.py:14:0: C0411: standard import "from types import TracebackType" should be placed before "from PIL import Image, ImageTk" (wrong-import-order)
src/guiguts/mainwindow.py:15:0: C0411: standard import "from typing import Any, Callable, Optional" should be placed before "from PIL import Image, ImageTk" (wrong-import-order)
************* Module guiguts.widgets
src/guiguts/widgets.py:114:0: C0301: Line too long (101/100) (line-too-long)
src/guiguts/widgets.py:75:4: W1113: Keyword argument before variable positional arguments list in the definition of __init__ function (keyword-arg-before-vararg)
src/guiguts/widgets.py:130:29: W0613: Unused argument 'event' (unused-argument)
src/guiguts/widgets.py:153:0: R0901: Too many ancestors (11/7) (too-many-ancestors)
src/guiguts/widgets.py:4:0: C0411: standard import "import tkinter as tk" should be placed before "import regex as re" (wrong-import-order)
src/guiguts/widgets.py:5:0: C0411: standard import "from tkinter import simpledialog, ttk" should be placed before "import regex as re" (wrong-import-order)
src/guiguts/widgets.py:6:0: C0411: standard import "from typing import Any, Optional, TypeVar" should be placed before "import regex as re" (wrong-import-order)
************* Module guiguts.application
src/guiguts/application.py:14:4: W0104: Statement seems to have no effect (pointless-statement)
src/guiguts/application.py:17:0: C0413: Import "from guiguts.file import File, NUM_RECENT_FILES" should be placed at the top of the module (wrong-import-position)
src/guiguts/application.py:18:0: C0413: Import "from guiguts.maintext import maintext" should be placed at the top of the module (wrong-import-position)
src/guiguts/application.py:19:0: C0413: Import "from guiguts.mainwindow import root, MainWindow, Menu, menubar, StatusBar, statusbar, ErrorHandler" should be placed at the top of the module (wrong-import-position)
src/guiguts/application.py:29:0: C0413: Import "from guiguts.page_details import PageDetailsDialog" should be placed at the top of the module (wrong-import-position)
src/guiguts/application.py:30:0: C0413: Import "from guiguts.preferences import preferences" should be placed at the top of the module (wrong-import-position)
src/guiguts/application.py:31:0: C0413: Import "from guiguts.preferences_dialog import PreferencesDialog" should be placed at the top of the module (wrong-import-position)
src/guiguts/application.py:32:0: C0413: Import "from guiguts.search import show_search_dialog, find_next" should be placed at the top of the module (wrong-import-position)
src/guiguts/application.py:33:0: C0413: Import "from guiguts.utilities import is_mac" should be placed at the top of the module (wrong-import-position)
src/guiguts/application.py:180:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/application.py:187:8: W0622: Redefining built-in 'help' (redefined-builtin)
src/guiguts/application.py:185:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/application.py:210:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/application.py:214:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/application.py:234:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/application.py:243:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/application.py:385:29: W0621: Redefining name 'statusbar' from outer scope (line 19) (redefined-outer-name)
src/guiguts/application.py:290:12: W0201: Attribute 'menu_file' defined outside __init__ (attribute-defined-outside-init)
src/guiguts/application.py:41:0: R0904: Too many public methods (32/20) (too-many-public-methods)
************* Module guiguts.file
src/guiguts/file.py:14:0: R0402: Use 'from guiguts import page_details' instead (consider-using-from-import)
src/guiguts/file.py:33:0: C0115: Missing class docstring (missing-class-docstring)
src/guiguts/file.py:128:12: W1203: Use lazy % formatting in logging functions (logging-fstring-interpolation)
src/guiguts/file.py:147:8: R1705: Unnecessary "else" after "return", remove the "else" and de-indent the code inside it (no-else-return)
src/guiguts/file.py:141:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/file.py:154:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/file.py:216:13: W1514: Using open without explicitly specifying an encoding (unspecified-encoding)
src/guiguts/file.py:236:11: W0621: Redefining name 'page_details' from outer scope (line 14) (redefined-outer-name)
src/guiguts/file.py:291:29: W0621: Redefining name 'page_details' from outer scope (line 14) (redefined-outer-name)
src/guiguts/file.py:314:32: W0621: Redefining name 'page_details' from outer scope (line 14) (redefined-outer-name)
src/guiguts/file.py:411:8: R1705: Unnecessary "else" after "return", remove the "else" and de-indent the code inside it (no-else-return)
src/guiguts/file.py:445:8: R1705: Unnecessary "else" after "return", remove the "else" and de-indent the code inside it (no-else-return)
src/guiguts/file.py:474:19: W0212: Access to a protected member _tkinter of a client class (protected-access)
src/guiguts/file.py:40:0: R0904: Too many public methods (33/20) (too-many-public-methods)
src/guiguts/file.py:8:0: C0411: standard import "import tkinter as tk" should be placed before "import regex as re" (wrong-import-order)
src/guiguts/file.py:9:0: C0411: standard import "from tkinter import filedialog, messagebox, simpledialog" should be placed before "import regex as re" (wrong-import-order)
src/guiguts/file.py:10:0: C0411: standard import "from typing import Any, Callable, Final, TypedDict, Literal, Optional" should be placed before "import regex as re" (wrong-import-order)
************* Module guiguts.preferences_dialog
src/guiguts/preferences_dialog.py:24:4: W0237: Parameter 'master' has been renamed to 'frame' in overriding 'PreferencesDialog.body' method (arguments-renamed)
************* Module guiguts.search
src/guiguts/search.py:33:4: R0914: Too many local variables (17/15) (too-many-locals)
src/guiguts/search.py:33:4: R0915: Too many statements (62/50) (too-many-statements)
src/guiguts/search.py:186:4: W1113: Keyword argument before variable positional arguments list in the definition of search_clicked function (keyword-arg-before-vararg)
src/guiguts/search.py:186:0: W0613: Unused argument 'args' (unused-argument)
src/guiguts/search.py:240:8: W0622: Redefining built-in 'range' (redefined-builtin)
src/guiguts/search.py:271:8: W0622: Redefining built-in 'range' (redefined-builtin)
************* Module guiguts.utilities
src/guiguts/utilities.py:14:0: C0103: Constant name "_called_from_test" doesn't conform to UPPER_CASE naming style (invalid-name)
src/guiguts/utilities.py:46:12: W0707: Consider explicitly re-raising using 'except AttributeError as exc' and 'raise Exception('Unknown windowing system') from exc' (raise-missing-from)
src/guiguts/utilities.py:46:12: W0719: Raising too general exception: Exception (broad-exception-raised)
src/guiguts/utilities.py:60:13: W1514: Using open without explicitly specifying an encoding (unspecified-encoding)
src/guiguts/utilities.py:64:16: W1201: Use lazy % formatting in logging functions (logging-not-lazy)
src/guiguts/utilities.py:107:0: R0903: Too few public methods (0/2) (too-few-public-methods)
src/guiguts/utilities.py:139:0: C0103: Constant name "_bell_callback" doesn't conform to UPPER_CASE naming style (invalid-name)
src/guiguts/utilities.py:147:4: W0603: Using the global statement (global-statement)
src/guiguts/utilities.py:8:0: C0411: standard import "from typing import Any, Optional, Callable" should be placed before "import regex as re" (wrong-import-order)
************* Module guiguts.__main__
src/guiguts/__main__.py:1:0: C0114: Missing module docstring (missing-module-docstring)
************* Module guiguts.page_details
src/guiguts/page_details.py:58:4: W0246: Useless parent or super() delegation in method '__init__' (useless-parent-delegation)
src/guiguts/page_details.py:66:12: W0612: Unused variable 'png' (unused-variable)
src/guiguts/page_details.py:127:4: W0237: Parameter 'master' has been renamed to 'frame' in overriding 'PageDetailsDialog.body' method (arguments-renamed)
src/guiguts/page_details.py:138:8: C0200: Consider using enumerate instead of iterating with range and len (consider-using-enumerate)
src/guiguts/page_details.py:4:0: C0411: standard import "import tkinter as tk" should be placed before "import roman" (wrong-import-order)
src/guiguts/page_details.py:5:0: C0411: standard import "from tkinter import simpledialog, ttk" should be placed before "import roman" (wrong-import-order)
src/guiguts/page_details.py:1:0: R0801: Similar lines in 2 files
==guiguts.maintext:[14:26]
==guiguts.mainwindow:[27:39]
logger = logging.getLogger(__package__)

TEXTIMAGE_WINDOW_ROW = 0
TEXTIMAGE_WINDOW_COL = 0
SEPARATOR_ROW = 1
SEPARATOR_COL = 0
STATUSBAR_ROW = 2
STATUSBAR_COL = 0
MIN_PANE_WIDTH = 20
TK_ANCHOR_MARK = "tk::anchor1"

 (duplicate-code)

------------------------------------------------------------------
Your code has been rated at 9.33/10 (previous run: 9.11/10, +0.23)
```

--- a/.github/workflows/python-app.yml
+++ b/.github/workflows/python-app.yml
@@ -30,6 +30,10 @@ jobs:
     - name: Lint with flake8
       run: |
         poetry run flake8 ./src
+    - name: Lint with pylint
+      run: |
+        poetry run pylint --recursive y .
+      continue-on-error: true
     - name: Check format with black
       run: |
         poetry run black --check --verbose ./src

--- a/README.md
+++ b/README.md
@@ -129,20 +129,29 @@ poetry env use $(brew --prefix)/bin/python3.11
 
 ## Code style
 
-Guiguts 2 uses [flake8](https://pypi.org/project/flake8) for static code analysis
-and [black](https://pypi.org/project/black) for consistent styling. Both use
-default settings, with the exception of maximum line length checking which is
-adjusted in the recommended manner using the `.flake8` file to avoid conflicts
-with black.
+Guiguts 2 uses [flake8](https://pypi.org/project/flake8) and
+[pylint](https://www.pylint.org) for static code analysis, and
+[black](https://pypi.org/project/black) for consistent styling.  All use default
+settings, with the exception of maximum line length checking which is adjusted
+in the recommended manner (using the `.flake8` file and the `tool.pylint`
+section of `pyproject.toml`) to avoid conflicts with black.
 
-Both tools will be installed via `poetry` as described above.
+All of the above tools will be installed via `poetry` as described above.
 
 `poetry run flake8 .` will check all `src` & `tests` python files.
 
+`poetry run pylint --recursive y .` will check all `src` & `tests` python files.
+
 `poetry run black .` will reformat all `src` & `tests` python files where necessary.
 
-This project uses Github Actions to ensure neither of the above tools reports any
+This project uses Github Actions to ensure neither flake8 nor black report any
 error.
+
+> **NOTE:** At this time, `pylint` is running in GitHub Actions in an advisory
+mode only.  That means that any warning or error reported by `pylint` will not
+cause the Action to fail. Over time, as we make changes in Guiguts and/or the
+`pylint` configuration, the aim is to clear all of `pylint`'s warnings. At that
+point perhaps it can run in an enforcement mode.
 
 Naming conventions from [PEP8](https://pep8.org/#prescriptive-naming-conventions)
 are used. To summarize, class names use CapWords; constants are ALL_UPPERCASE;

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -18,6 +18,7 @@ flake8 = "^6.1"
 Sphinx = "^7.2"
 pytest = "^7.4"
 mypy = "^1.8"
+pylint = "^3.0.3"
 types-Pillow = "^10.1"
 types-regex = "^2023.12"
 build = "^1.0.3"
@@ -28,3 +29,21 @@ guiguts = "guiguts.application:main"
 [build-system]
 requires = ["poetry-core"]
 build-backend = "poetry.core.masonry.api"
+
+[tool.pylint."messages control"]
+confidence = ["HIGH", "CONTROL_FLOW", "INFERENCE", "INFERENCE_FAILURE", "UNDEFINED"]
+
+disable = [
+  "raw-checker-failed",
+  "bad-inline-option",
+  "locally-disabled",
+  "file-ignored",
+  "suppressed-message",
+  "useless-suppression",
+  "deprecated-pragma",
+  "use-symbolic-message-instead",
+  "line-too-long"
+]
+
+enable = ["c-extension-no-member"]
+

